### DISCUSSION TITLE
What am I doing wrong in this org-ql-block query using parent or ancestors?

### DISCUSSION THREAD

**Discussion by al3xandru** [alphapapa/org-ql#410] [Q&A] [answered]
The following 2 queries are giving me surprising results making me realize that I don't fully understand how they work.

```
(org-ql-block '(parent (heading "[INBOX]")))

(org-ql-block '(and
                 (todo "TODO" "WIP")
                 (tags "work")
                 (ancestors (heading "[INBOX]"))))
```

The structure of the org file is:

```org
* [INBOX]
** TODO Task a
** TODO Work task A :work:
* [NOTES]
* 2024
** 2024-W08
```

The first returns every item in the file.
The second returns every TODO/WIP tagged with work across the file.

It looks like my usage of `parent` and `ancestors` is not correct. I've read the docs and searched for similar examples.  I'm probably missing something or skipped an important element as I have 2 more usages of ancestors that are working well (the only exception being that they are deeper in the outline structure).

Could anyone point me into the right direction? thank you

**Comment by alphapapa** [comment]
Hi Alex,

Do you get the same incorrect results when using `org-ql-search` instead of `org-ql-block`?

**Comment by al3xandru** [comment]
1. Yes.
2. It seems there is an issue that starts with `(heading "[INBOX]")` which instead of returning a single result it returns 4 items.
3. The other 3 results are children of "[INBOX]" and thus would not explain the query going outside that parent.
4. What bothers me is that the exact type of queries return the results I expect when matching deeper level items

**Comment by al3xandru** [comment]
Taking a hint from point 4 above, I have moved the INBOX item and all its children to a lower level.  The `parent` & `ancestors` based queries returned the correct results (well, the results I expected at least)

My understanding of `heading` predicate is that it matches if the heading text matching the string(s).  Is that correct? I can't find a good explanation why there are 4 results for the `(heading "[INBOX]")` query (the parent heading is not prepended to the child, right? otherwise all children would match)

**Comment by alphapapa** [comment]
Well, I don't understand what's going on here, because I wasn't able to reproduce the problem.  Here's what I did:

1.  Using https://github.com/alphapapa/with-emacs.sh, run `with-emacs.sh -i org-ql`.
2. Make file `/tmp/test.org`, paste your contents into it, and save it.
3. Run `org-ql-search` on that file with each of the queries you gave.  They both returned the correct, expected results, without any headings that didn't belong.

I'd suggest trying to reproduce the problem with those steps.  Then be sure to mention what versions of Emacs, Org, and org-ql you're using.

**Comment by al3xandru** [comment]
1. I have tried to use `with-emacs.sh` and it complains about unsupported `o` option.
2. In the with-emacs version, I have installed org-ql using `package-install`, and then required it
3. Copied over my org file. Then chopped it down to a short version that still returns what I consider invalid results:

```
#+TITLE: my life organized
* [INBOX]
:PROPERTIES:
:ID:       967A3D3F-D8B9-487B-9803-60BAC1F9853A
:END:
** TODO Subtask 1
** TODO Subtask 2
* 2024
:PROPERTIES:
:ID:       0FCB3663-B044-4BF1-8A3C-CDAA3F2C61F0
:END:
** 2024-W08
:PROPERTIES:
:ID:       04B8D416-24EA-4710-976C-ABF60F7BA317
:END:
*** [#B] Weekly Program [52%]
**** TODO Power over
*** 2024-02-20 Tuesday
***** DONE Start work 
SCHEDULED: <2024-02-20 Tue 17:00>
***** DONE Do Good Work
CLOSED: [2024-02-20 Tue 17:20]
****** DONE Estimate available work time (1h30ʹ)
CLOSED: [2024-02-20 Tue 17:12]
21:45-17:15=4h30ʹ-1h30ʹ=3h
- 1.5h Meetings
```

4. executed the query `(ancestors (heading "[INBOX]"))` which returns 4 results instead of 2

![image](https://github.com/alphapapa/org-ql/assets/290841/f13080e3-0185-4a92-a36c-6be48c8f23b2)

**Comment by al3xandru** [comment]
As regards the `with-emacs.sh` does this command look ok to you:

```
Running: emacs --quick --title Emacs (config: /var/folders/f4/5nbbgd055px6mv0hftysx5cw0000gn/T/tmp.KYt0p3Cj8J) --eval (setq user-emacs-directory (file-truename "/var/folders/f4/5nbbgd055px6mv0hftysx5cw0000gn/T/tmp.KYt0p3Cj8J")) --eval (setq user-init-file (expand-file-name "init.el" user-emacs-directory)) -l package --eval (when (boundp 'native-comp-eln-load-path) (push (expand-file-name "eln-cache" user-emacs-directory) native-comp-eln-load-path)) --eval (add-to-list 'package-archives '("gnu" . "https://elpa.gnu.org/packages/") t) --eval (add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t) --eval (add-to-list 'package-archives '("org" . "https://orgmode.org/elpa/") t) --eval (setq package-user-dir (expand-file-name "elpa" user-emacs-directory)) --eval (package-initialize) --eval (package-refresh-contents) --eval (when (file-exists-p user-init-file) (load-file user-init-file)) ./with-emacs.sh -o d:e:hi:OPR -l dir:,debug,emacs:,help,install:,no-package,no-org-repo,no-refresh-packages -- -i org-ql
```

**Comment by alphapapa** [comment]
This appears to be https://github.com/alphapapa/org-ql/issues/365

**Comment by al3xandru** [comment]
Ha... I did not pick up 365 when doing a search prior to posting.

Any suggestions related to with-emacs? It's an interesting utility that could prove useful.  I'm on a macOS environment.

**Reply by alphapapa** [reply]
> Ha... I did not pick up 365 when doing a search prior to posting.

And a quick check does not reveal the cause of the problem.  It appears that `org-ql--value-at` is returning an incorrect non-nil value for the parent node of the nodes that should not be in the result, but I don't know why.

I guessed you are on a Mac.  You will need to install GNU userland (coreutils, or whatever it's called in Homebrew) and make sure your PATH is configured so the GNU commands come first in your environment.  Please consult other Mac users.

**Comment by alphapapa** [comment, marked as answer]
> This appears to be #365

Correction, this appears to have been a bug in the `heading` predicate, which just happened to be noticed in conjunction with the `parent` predicate here.  I'll be releasing a fix in a v0.8.5 release.

**Reply by alphapapa** [reply]
See also new release which fixes `ancestors` and `parent`: https://github.com/alphapapa/org-ql/releases/tag/v0.8.5

**Comment by al3xandru** [comment]
If anyone reads this discussion again and wonders about `with-emacs` on macOS:

1. if running brew, then  `brew install gnu-getopt`
2. macOS has `getopt` in `/usr/bin/getopt` thus, put on the gnu-opt bin dir first in the PATH (e.g. `/opt/homebrew/opt/gnu-getopt/bin`) before running with-emacs.sh

(@alphapapa if you think this is useful for other with-emacs potential users, please do use it as you see fit)

**Reply by alphapapa** [reply]
Thanks.  It's also discussed at https://github.com/alphapapa/with-emacs.sh/issues/3, which links to a similar discussion.
